src/Common/ThreadPool.cpp

```diff
@@ -238,6 +238,8 @@ void ThreadPoolImpl<Thread>::worker(typename std::list<Thread>::iterator thread_
             }
             catch (...)
             {
+                job = {};
```

Comment:
What race exactly does this line prevent? It needs a comment in the code.

In the gist you posted (https://gist.github.com/azat/d480dc3af5a0a44de4b038e20807c4b9) the race seems to be between server termination and some worker thread. It is due to not waiting properly for the pool shutdown, but looks like it should be fixed by the `wait()` call, and this line fixes something else?

Reply:
>What race exactly does this line prevent?

w/o reseting the Job the waiter can return earlier then it will be destroyed, since it waits only when scheduled_jobs goes to zero (this is a follow up for #15061)

> It needs a comment in the code.

Ok